src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -72,17 +68,85 @@ public HttpConnectionPools(HttpConnectionSettings settings, bool usingProxy)
         }
 
         public HttpConnectionSettings Settings => _settings;
-        public bool UsingProxy => _usingProxy;
 
-        /// <summary>Gets a pool for the specified endpoint, adding one if none existed.</summary>
-        /// <param name="key">The endpoint for the pool.</param>
-        /// <returns>The retrieved pool.</returns>
-        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key)
+        private static string ParseHostNameFromHeader(string hostHeader)
+        {
+            // See if we need to trim off a port.
+            int colonPos = hostHeader.IndexOf(':');
+            if (colonPos >= 0)
+            {
+                // There is colon, which could either be a port separator or a separator in
+                // an IPv6 address.  See if this is an IPv6 address; if it's not, use everything
+                // before the colon as the host name, and if it is, use everything before the last
+                // colon iff the last colon is after the end of the IPv6 address (otherwise it's a
+                // part of the address).
+                int ipV6AddressEnd = hostHeader.IndexOf(']');
+                if (ipV6AddressEnd == -1)
+                {
+                    return hostHeader.Substring(0, colonPos);
+                }
+                else
+                {
+                    colonPos = hostHeader.LastIndexOf(':');
+                    if (colonPos > ipV6AddressEnd)
+                    {
+                        return hostHeader.Substring(0, colonPos);
+                    }
+                }
+            }
+
+            return hostHeader;
+        }
+
+        private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Uri proxyUri)
         {
+            Uri uri = request.RequestUri;
+
+            string sslHostName = null;
+            if (HttpUtilities.IsSupportedSecureScheme(uri.Scheme))
+            {
+                string hostHeader = request.Headers.Host;
+                if (hostHeader == null)
+                {
+                    // No explicit Host header.  Use host from uri.
+                    sslHostName = uri.IdnHost;
+                }
+                else
+                {
+                    sslHostName = ParseHostNameFromHeader(hostHeader);
+                }
```

Comment:
Minor personal preference, I find constructs like this easier to understand when the polarity is inverted; just less for me to remember as I'm reading the code, i.e.
"if the host header isn't null, then parse it, otherwise use idnhost"
vs
"if the host header is null, then use idnhost, otherwise parse the host header"

Reply:
I generally agree.  On the other hand, when there is a common case and an atypical case (as here) I like to make the common case come first, i.e. if case rather than else.

Reply:
I went ahead and switched this.

Reply:
Thanks :)